RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h

```diff
@@ -0,0 +1,184 @@
+#ifndef RecoEgamma_EgammaTools_EnergyScaleCorrection_h
+#define RecoEgamma_EgammaTools_EnergyScaleCorrection_h
+
+//author: Alan Smithee
+//description: 
+//  A port of Shervin Nourbakhsh's EnergyScaleCorrection_class in EgammaAnalysis/ElectronTools
+//  this reads the scale & smearing corrections in from a text file for given categories
+//  it then allows these values to be accessed
+
+
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <cmath>
+#include <string>
+#include <bitset> 
+
+
+class EnergyScaleCorrection
+{	
+public:  
+  enum FileFormat{
+    UNKNOWN=0,
+    GLOBE,
+    ECALELF_TOY,
+    ECALELF
+  };
+
+  enum ParamSmear{
+    kNone = 0,
+    kRho,
+    kPhi,
+    kNParamSmear
+  };
+  
+  enum ScaleNuisances{
+    kErrStatBitNr = 0,
+    kErrSystBitNr = 1,
+    kErrGainBitNr = 2,
+    kErrNrBits=3,
+    kErrNone = 0,
+    kErrStat = 1,
+    kErrSyst = 2,
+    kErrGain = 4,
+    kErrStatSyst = 3,
+    kErrStatGain = 5,
+    kErrSystGain = 6,
+    kErrStatSystGain = 7
+  };
+  
+  
+
+  
+  class ScaleCorrection
+  {
+  public:  
+    ScaleCorrection():
+      scale_(1.),scaleErrStat_(0.),scaleErrSyst_(0.),scaleErrGain_(0.){}
+    ScaleCorrection(float iScale,float iScaleErrStat,float iScaleErrSyst,float iScaleErrGain):
+      scale_(iScale),scaleErrStat_(iScaleErrStat),scaleErrSyst_(iScaleErrSyst),scaleErrGain_(iScaleErrGain){}
+    
+    float scale()const{return scale_;}
+    float scaleErr(const std::bitset<kErrNrBits>& uncBitMask)const;
+    float scaleErrStat()const{return scaleErrStat_;}
+    float scaleErrSyst()const{return scaleErrSyst_;}
+    float scaleErrGain()const{return scaleErrGain_;}
+
+    friend std::ostream& operator<<(std::ostream& os, const ScaleCorrection& a){return a.print(os);}
```

Comment:
why does this need to be a friend? it calls a public method

Reply:
simply a way to be able to define the operator in the body of the class so its obvious its available.

Reply:
This is a bit of an abuse of the "friend" feature of the language.